File: solutions/python/replace_and_remove.py
import copy
import sys
import random


# @include
def replace_and_remove(size, s):
    # Forward iteration: remove 'b's and count the number of 'a's.
    write_idx, a_count = 0, 0
    for i in range(size):
        if s[i] != 'b':
            s[write_idx] = s[i]
            write_idx += 1
        if s[i] == 'a':
            a_count += 1

    # Backward iteration: replace 'a's with 'dd's starting from the end.
    cur_idx = write_idx - 1
    write_idx += a_count - 1
    final_size = write_idx + 1
    while cur_idx >= 0:
        if s[cur_idx] == 'a':
            s[write_idx - 1:write_idx + 1] = 'dd'
            write_idx -= 2
        else:
            s[write_idx] = s[cur_idx]
            write_idx -= 1
        cur_idx -= 1
    return final_size
# @exclude


def rand_string(length):
    return [random.choice('abcd') for _ in range(length)]


def check_ans(s, ans):
    temp = []
    for i in s:
        if i == 'a':
            temp += ['d', 'd']
        elif i != 'b':
            temp.append(i)
    assert ans == temp


def main():
    for _ in range(1000):
        if len(sys.argv) == 2:
            s = list(sys.argv[1])
        else:
            s = rand_string(random.randint(1, 1000))
        print(''.join(s))
        s_copy = copy.deepcopy(s)
        s_copy += [0] * len(s)
        final_size = replace_and_remove(len(s), s_copy)
        check_ans(s, s_copy[:final_size])


if __name__ == '__main__':
    main()

File: solutions/java/src/main/java/com/epi/ReplaceAndRemove.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
package com.epi;

import java.util.Arrays;
import java.util.Random;

public class ReplaceAndRemove {
  // @include
  public static String replaceAndRemove(String s) {
    char[] sChars = s.toCharArray();
    // Forward iteration: remove "b"s and count the number of "a"s.
    int writeIdx = 0, aCount = 0;
    for (char c : sChars) {
      if (c != 'b') {
        sChars[writeIdx++] = c;
      }
      if (c == 'a') {
        ++aCount;
      }
    }

    // Allocates space according to the number of "a".
    sChars = Arrays.copyOf(sChars, writeIdx + aCount);
    // Backward iteration: replace "a"s with "dd"s starting from the end.
    int curIdx = writeIdx - 1;
    writeIdx = sChars.length - 1;
    while (curIdx >= 0) {
      if (sChars[curIdx] == 'a') {
        sChars[writeIdx--] = 'd';
        sChars[writeIdx--] = 'd';
      } else {
        sChars[writeIdx--] = sChars[curIdx];
      }
      --curIdx;
    }
    return new String(sChars);
  }
  // @exclude

  static String randString(int len) {
    Random gen = new Random();
    StringBuilder ret = new StringBuilder();
    while (len-- > 0) {
      ret.append((char)(gen.nextInt(4) + 97));
    }

    return ret.toString();
  }

  static void checkAns(String s, String ans) {
    StringBuilder temp = new StringBuilder();
    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == 'a') {
        temp.append('d');
        temp.append('d');
      } else if (s.charAt(i) != 'b') {
        temp.append(s.charAt(i));
      }
    }
    assert(ans.equals(temp.toString()));
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      String s;
      if (args.length == 1) {
        s = args[0];
      } else {
        s = randString(gen.nextInt(1000) + 1);
      }
      System.out.println(s);
      System.out.println();
      String ans = replaceAndRemove(s);
      System.out.println(ans);
      checkAns(s, ans);
    }
  }
}
